### PR TITLE
feat(goal_planner): calculate stop pose from closest goal candidate

### PR DESCRIPTION
**Pull request by kosuke55**
## Description

currently the stop pose when path is not found is caculated from goal search start pose.
this include invalid goal pose like in no_parking_lane and stop too far.
In this PR, calculate stop pose from actual goal candidates which do not include no_parking_lane goal.

before

image

after

image

## Related links

this may solve degradation of https://github.com/autowarefoundation/autoware.universe/pull/5371#issuecomment-1776459751

## Tests performed

psim

evaluator_description: feat/goal_planner_stop_pose
2023/10/25 https://evaluation.tier4.jp/evaluation/reports/8bfda284-da50-5c03-ac67-a2f834d6c731/?project_id=prd_jt

evaluator_description: feat/goal_planner_stop_pose
2023/10/26 https://evaluation.tier4.jp/evaluation/reports/04de2c84-3b7f-5299-b0b1-9eb0d98f6721/?project_id=prd_jt

## Interface changes

none

## Effects on system behavior

none

## Pre-review checklist for the PR author

The PR author **must** check the checkboxes below when creating the PR.

- [x] I've confirmed the [contribution guidelines].
- [x] The PR follows the [pull request guidelines].

## In-review checklist for the PR reviewers

The PR reviewers **must** check the checkboxes below before approval.

- [ ] The PR follows the [pull request guidelines].
- [ ] The PR has been properly tested.
- [ ] The PR has been reviewed by the code owners.

## Post-review checklist for the PR author

The PR author **must** check the checkboxes below before merging.

- [ ] There are no open discussions or they are tracked via tickets.
- [ ] The PR is ready for merge.

After all checkboxes are checked, anyone who has write access can merge the PR.

[contribution guidelines]: https://autowarefoundation.github.io/autoware-documentation/main/contributing/
[pull request guidelines]: https://autowarefoundation.github.io/autoware-documentation/main/contributing/pull-request-guidelines/

### CODE REVIEW

File: planning/behavior_path_planner/src/utils/goal_planner/goal_searcher.cpp [hunk 1]
```diff
@@ -468,4 +468,28 @@ bool GoalSearcher::isInAreas(const LinearRing2d & footprint, const BasicPolygons
...
+    /*forward_only_in_route*/ false);
+
+  const auto closest_goal_candidate = std::min_element(
+    goal_candidates.begin(), goal_candidates.end(),
+    [&current_lanes](const auto & a, const auto & b) {
+      const auto goal_arc_length_a =
+        lanelet::utils::getArcCoordinates(current_lanes, a.goal_pose).length;
+      const auto goal_arc_length_b =
+        lanelet::utils::getArcCoordinates(current_lanes, b.goal_pose).length;
+      return goal_arc_length_a < goal_arc_length_b;
```
Comment (kyoichi-sugahara) [inline planning/behavior_path_planner/src/utils/goal_planner/goal_searcher.cpp:485]: I just prefer this way personally
```suggestion
  // Define a lambda function to compute the arc length for a given goal candidate.
  auto getGoalArcLength = [&current_lanes](const auto & candidate) {
    return lanelet::utils::getArcCoordinates(current_lanes, candidate.goal_pose).length;
  };

  // Find the closest goal candidate by comparing the arc lengths of each candidate.
  const auto closest_goal_candidate = std::min_element(
    goal_candidates.begin(), goal_candidates.end(),
    [&getGoalArcLength](const auto & a, const auto & b) {
      return getGoalArcLength(a) < getGoalArcLength(b);
    });
```

**Review by kyoichi-sugahara — APPROVED** [review]
LGTM!

File: planning/behavior_path_planner/src/utils/goal_planner/goal_searcher.cpp (hunk 1 above)
Comment (kosuke55) [reply planning/behavior_path_planner/src/utils/goal_planner/goal_searcher.cpp:485]: seems better!!

File: planning/behavior_path_planner/src/utils/goal_planner/goal_searcher.cpp [hunk 2]
```diff
@@ -468,4 +468,31 @@ bool GoalSearcher::isInAreas(const LinearRing2d & footprint, const BasicPolygons
...
+    return lanelet::utils::getArcCoordinates(current_lanes, candidate.goal_pose).length;
+  };
+
+  // Find the closest goal candidate by comparing the arc lengths of each candidate.
+  const auto closest_goal_candidate = std::min_element(
+    goal_candidates.begin(), goal_candidates.end(),
+    [&getGoalArcLength](const auto & a, const auto & b) {
+      return getGoalArcLength(a) < getGoalArcLength(b);
+    });
+});
```
Comment (kosuke55) [inline planning/behavior_path_planner/src/utils/goal_planner/goal_searcher.cpp:489]:
```suggestion
```

File: planning/behavior_path_planner/src/scene_module/goal_planner/goal_planner_module.cpp [hunk 3]
```diff
@@ -547,13 +547,19 @@ void GoalPlannerModule::generateGoalCandidates()
     goal_searcher_->setPlannerData(planner_data_);
     goal_searcher_->setReferenceGoal(status_.get_refined_goal_pose());
     status_.set_goal_candidates(goal_searcher_->search());
+    const auto current_lanes = utils::getExtendedCurrentLanes(
+      planner_data_, parameters_->backward_goal_search_length,
+      parameters_->forward_goal_search_length, false);
```
Comment (kosuke55) [inline planning/behavior_path_planner/src/scene_module/goal_planner/goal_planner_module.cpp:552]:
```suggestion
```

**Pull request merged**